parsel/selector.py

```diff
@@ -285,29 +330,50 @@ def __init__(
         text: Optional[str] = None,
         type: Optional[str] = None,
         namespaces: Optional[Mapping[str, str]] = None,
-        root: Optional[Any] = None,
+        root: Optional[Any] = _NOTSET,
         base_url: Optional[str] = None,
         _expr: Optional[str] = None,
         huge_tree: bool = LXML_SUPPORTS_HUGE_TREE,
     ) -> None:
-        self.type = st = _st(type or self._default_type)
-        self._parser = _ctgroup[st]["_parser"]
-        self._csstranslator = _ctgroup[st]["_csstranslator"]
-        self._tostring_method = _ctgroup[st]["_tostring_method"]
+        self.root: Any
+        if type not in ("html", "json", "text", "xml", None):
+            raise ValueError(f"Invalid type: {type}")
 
-        if text is not None:
-            if not isinstance(text, str):
-                msg = f"text argument should be of type str, got {text.__class__}"
-                raise TypeError(msg)
-            root = self._get_root(text, base_url, huge_tree)
-        elif root is None:
+        if text is None and root is _NOTSET:
             raise ValueError("Selector needs either text or root argument")
 
+        if text is not None and not isinstance(text, str):
+            msg = f"text argument should be of type str, got {text.__class__}"
+            raise TypeError(msg)
+
+        if text is not None:
+            if root is not _NOTSET:
+                warnings.warn(
+                    "Selector got both text and root, root is being ignored.",
+                    stacklevel=2,
+                )
+            if type == "text":
+                self.type = type
+                self.root = text
+            elif (
+                type == "json" or bool(_load_json_or_none(text))
+            ) and not isinstance(_load_json_or_none(text), str):
+                self.type = "json"
+                self.root = _load_json_or_none(text)
```

Comment:
Could you please try to optimize it a bit more, so that JSON is not loaded multiple times? `_is_valid_json` may also be problematic, but it may be harder to optimize.

Reply:
without walrus operator the only two options are to use the method before the for loop or to remove the `json.loads`  from it and renaming it to `is_valid_json`. It may be that just an `is_valid_json` won't cover all cases though.